src/metadatad.cc

```diff
@@ -14,11 +14,59 @@
  * limitations under the License.
  **/
 
+#include <csignal>
+#include <cstdlib>
+#include <initializer_list>
+#include <iostream>
+
 #include "agent.h"
 #include "configuration.h"
 #include "docker.h"
 #include "instance.h"
 #include "kubernetes.h"
+#include "time.h"
+
+namespace google {
+namespace {
+
+class CleanupState {
+ public:
+  CleanupState(
+      std::initializer_list<MetadataUpdater*> updaters, MetadataAgent* server)
+      : updaters_(updaters), server_(server) { server_wait_mutex_.lock(); }
+
+  void StartShutdown() const {
+    std::cerr << "Stopping server" << std::endl;
+    server_->Stop();
+    std::cerr << "Stopping updaters" << std::endl;
+    for (MetadataUpdater* updater : updaters_) {
+      updater->NotifyStop();
+    }
+    server_wait_mutex_.unlock();
+    // Give the notifications some time to propagate.
+    std::this_thread::sleep_for(time::seconds(0.1));
+  }
+
+  void Wait() const {
+    std::lock_guard<std::mutex> await_server_shutdown(server_wait_mutex_);
+  }
+
+ private:
+  mutable std::mutex server_wait_mutex_;
+  std::vector<MetadataUpdater*> updaters_;
+  MetadataAgent* server_;
+};
+const CleanupState* cleanup_state;
+
+}  // namespace
+}  // google
+
+extern "C" [[noreturn]] void handle_sigterm(int signum) {
+  std::cerr << "Caught SIGTERM; shutting down" << std::endl;
+  google::cleanup_state->StartShutdown();
+  std::cerr << "Exiting" << std::endl;
+  std::exit(128 + signum);
```

Comment:
Exit code should be 0 if we terminate successfully. If it's anything but 0, Kubernetes will think that the pod crashed.

Reply:
This will only happen when a pod is killed by a health check. Do we really want to report a successful exit in that case? I seem to recall that there was a distinction in pod restart behavior between success and failure exits...

Reply:
That's a [`fluentd` thing](https://github.com/fluent/fluentd/blob/v0.12/lib/fluent/supervisor.rb#L537). Any other process will exit with exactly this exit code (i.e., 143) on `SIGTERM`. That's just a Unix convention. I bet heapster would exit with 143 as well.

Reply:
We spoke offline, I believe we're going to go with a 0 exit status code on a healthy exit after looking further into how it could positively impact docker.